Refactoring an anonymous class to inner class #SCL-227

leonbur opened the pull request:
This commit fixes this issue

https://youtrack.jetbrains.com/issue/SCL-227

**UPD**
Related thread in the discord 
https://discord.com/channels/931170831139217469/931171260824707072/1198202206260834344

azdrojowa123 committed:
sbt: do not use project to reset the model in SbtProjectSettingsControl.scala #SCL-21994 fixed

azdrojowa123 committed:
sbt: do not use project to reset the model in SbtProjectSettingsControl.scala #SCL-21994 fixed

SrTobi committed:
[lexer] fixed discontinuous sequence of tokens when newline is escaped #SCL-22003 fixed

Commit:
[refactoring] anonymous class to inner class #SCL-227

unkarjedy commented on scala/scala-impl/resources/messages/ScalaBundle.properties:
```diff
@@ -1804,6 +1804,7 @@ move.members=Move members
 move.members.supported.only.stable.objects=Move refactoring is supported only for stable object members
 move.members.not.supported.implicits=Move refactoring is not supported for implicit definitions
 move.members.not.supported.overridden=Move refactoring is not supported for overridden definitions
+move.anonymousToInner.name=Convert Anonymous to Inner...
```
In the dialog title it should not have "...", like in Java alternative.
https://jetbrains.design/intellij/text/punctuation/#ellipsis

leonbur replied:
from your link:
>End an action name with an ellipsis if it opens a dialog where input is required or possible. An ellipsis helps users understand if an action is immediate or additional interaction is to follow.

but the action does open a dialog and input is required. also, this action comes as a replacement to "Move class..." which also has ellipsis.

unkarjedy replied:
The action does open a dialog, but in the dialog itself also has a title and it shouldn't have the `...`.
See, for example `JavaRefactoringBundle.properties` - it has two keys:
```
anonymousToInner.refactoring.name=Convert Anonymous to Inner
...
convert.anonymous.to.inner.action.name=Convert Anonymous to Inner\u2026
```

One is used in the action and another in the title.
`\u2026` ~ UTF-8 for `...`

unkarjedy commented:
The refactoring should be able to handle closures.
For a start, it would be enough to at least handle read operations.
Ideally, write operations could be also handled.

For example, in this case the refactoring breaks the code:
```scala
trait MyTrait {
  def foo0: Int
  def foo1: Int
  def foo2: Int
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar1 = 42

    def innerFoo(): Unit = {
      var localVar2 = 42
      
      val x = new MyTrait {
        override def foo0: Int = args.length
        override def foo1: Int = localVar1
        override def foo2: Int = localVar2
      }
    }
  }
}
```
<img width="465" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/e21dd4fc-0ce7-4a29-b5ff-5e2cab14afd4">

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+class ScalaAnonymousToInnerHandler {
```
I see at least two issues with the current implementation:
1. It does not handle local val/var (see `org.jetbrains.plugins.scala.lang.psi.api.statements.ScValueOrVariable`).
I think for a start, we can just handle read operations. Handling write operations to `var` could be handled in theory, but that would require a more complex design and implementation.

2. It captures class parameters of the class it extracts the class (`classParameter`, `classParameterField`).

Example (1 + 2)
```scala
trait MyTrait {
  def traitFoo: Int
}

class Example(classParameter: Int, val classParameterField: Int) {
  val fieldVal: Int = 0
  val fieldVar: Int = 0

  def foo(functionParameter: Int): Unit = {
    val localValue = 0
    var localVariable = 0

    def innerFoo(innerFunctionParameter: Int): Unit = {
      val localValueInner = 0
      var localVariableInner = 0

      val x = new MyTrait {
        override def traitFoo: Int = {
          //writes (might skip in the initial implementation)
          localVariable = 42
          localVariableInner = 23

          //reads
          classParameter + classParameterField +
            fieldVal + fieldVar +
            functionParameter + innerFunctionParameter +
            localValue + localVariable +
            localValueInner + localVariableInner
        }
      }
    }
  }
}
```

<img width="1157" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/52ee7e80-e019-4fc0-9016-f36a25df2ddf">

<img width="751" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/8b606ee3-87bb-46d1-8b80-7127b84cf3c0">


Example in Java (for 2)
```java
interface MyInterface {
    int foo();
}

class Scratch {
    public int fieldPublic = 0;
    public int fieldPrivate = 0;

    public void foo(int parameter) {
        int localVar = 0;
        System.out.println(new MyInterface() {
            @Override
            public int foo() {
                return parameter + localVar + fieldPublic + fieldPrivate;
            }
        });
    }
}
```

<img width="685" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/3f79cfa1-8f8c-4f0d-8a38-fb22932239ab">

Note that in Java implementation it shows an extra option to create "static" class if there are no usages of class fields.
In future, in Scala refactoring, we could add an option to create the class in the companion object.
In that case it would capture class fields/parameters as well.
But it's fine if we don't have it in the first version.

leonbur replied:
My solution is basically to add the referenced vars and declare them as `var` in the parameter declaration of the class so that they could be mutated inside the newly created class.

unkarjedy replied:
@leonbur 
Oh, sorry, I missed this update.
> My solution is basically to add the referenced vars and declare them as var in the parameter declaration of the class so that they could be mutated inside the newly created class.

Unfortunately, this is not a correct solution.
Such refactoring breaks the semantics of the code, which is definitely not expected from the "refactoring"

For example, suppose you have this code:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0
    val t = new MyTrait {
      override def foo(): Unit = {
        localVar += 1
      }
    }
    t.foo()
    t.foo()
    println(localVar)
  }
}
```
it outputs `2`

After you apply the refactoring the code becomes as follows:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0
    val t = new MyMyTrait(localVar)
    t.foo()
    t.foo()
    println(localVar)
  }

  class MyMyTrait(var localVar: Int) extends MyTrait {
    override def foo(): Unit = {
      localVar += 1
    }
  }
}
```
and the output is `0`, because the original `var` is not changed.

unkarjedy replied:
When I wrote
> Handling write operations to var could be handled in theory, but that would require a more complex design and implementation.

I kept in mind another solution.
You would need to create a `Ref` wrapper for the var.
That's what actually Scala compiler automatically does in the JVM byte code.
This is the decompiled version:
<img width="607" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/8dfd8802-e7be-49c3-b46b-936af5fba7ed">

However I am not 100% sure that we should do the same.
Users might want to address this issue in some different way.
We might leave `localVar` an ordinary parameter.
The resulting code won't compile in the places where `localVar` is updated, but it's better than silently breaking the code.
We can add some special line comment in every `localVar =`, something like
`// refactor your code or use scala.runtime.ObjectRef`
(we the actual infer `*Ref` name from the variable type)

Ideally, we could add some setting in the dialog in order user can do it automatically.
But the UX needs more thinking.
For this first version it would be OK to do as described above.

unkarjedy replied:
Also, this part is still unresolved:
> It captures class parameters of the class it extracts the class (classParameter, classParameterField).
>...
>Note that in Java implementation it shows an extra option to create "static" class if there are no usages of class fields.
In future, in Scala refactoring, we could add an option to create the class in the companion object.
In that case it would capture class fields/parameters as well.
But it's fine if we don't have it in the first version.

<img width="611" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/2248167d-3b5d-47d4-ba1d-8a010e42f907">

leonbur replied:
ok, now understand. I'll add a comment and remove the `var` prefix for writable parameters.

regarding fieldVal,classParameter,... do you want them to NOT become parameters of the created class?

unkarjedy replied:
> regarding fieldVal,classParameter,... do you want them to NOT become parameters of the created class?

Yes, because those will be accessible in the extracted inner class automatically.

unkarjedy replied:
The new line is added in the wrong place for the comment
<img width="451" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/8dacd35e-95c9-4e04-9458-7de11a748854">

Other thoughts:
1. Let's prepend "todo"
2. Let's mention the var names in order it's clear what that comment is about 

Something like
```scala
//todo: refactor your code or use scala.runtime.ObjectRef (localVar)
//todo: refactor your code or use scala.runtime.ObjectRef (localVar1, localVar2)
```

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerDialog.scala:
```diff
@@ -0,0 +1,116 @@
+import javax.swing.{JComponent, JPanel}
+import scala.jdk.CollectionConverters._
+
+class ScalaAnonymousToInnerDialog(project: Project, anonClass: ScExtendsBlock, variables: Array[VariableData], target: Either[ScFile, ScTemplateDefinition]) extends DialogWrapper(project) {
```
Am I right that the original implementation was inspired by `com.intellij.refactoring.anonymousToInner.AnonymousToInnerDialog`?

If so, let's add a ScalaDoc comment:
```scala
/**
 * @note original implementation was inspired by [[com.intellij.refactoring.anonymousToInner.AnonymousToInnerDialog]]
 */
```

It might help in future in case we want to adapt some new improvements from the Java version.

unkarjedy replied:
The same is actual for other classes if applicable

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+
+  }, getRefactoringName, null)
+
+  @tailrec
```
We already have some utility extension methods for that.
Please see `org.jetbrains.plugins.scala.extensions.PsiElementExt#parentOfType`.

leonbur replied:
lol, I even used it elsewhere. will amend.

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            val containingClass = PsiTreeUtil.getParentOfType(p, classOf[ScClass])
```
Did you also consider `ScTrait`, `ScEnum` in addition to `ScClass` and `ScObject`?

leonbur replied:
I'm removing the parent check as I don't even remember why I added it (I started work on this a few months ago). it works even without it.

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+
+  private def showRefactoringDialog(project: Project, anonClass: ScExtendsBlock, usedVariables: Array[VariableData], target: Either[ScFile, ScTemplateDefinition]) = {
+    val dialog = new ScalaAnonymousToInnerDialog(project, anonClass, usedVariables, target)
+    if (dialog.showAndGet()) {
```
(minor) might omit wrapping braces `{}` here (as you did on the next line with `None`)

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  @tailrec
+  private def findAnonymousClass(element: PsiElement): Option[ScExtendsBlock] =
```
The naming is confusing.
I would expect findAnonymous**Class** to return a subtype of "class" (ScNewTemplateDefinition or PsiClass).
Let's replace this with `element.parentOfType[ScNewTemplateDefinition]`.
(requires `import org.jetbrains.plugins.scala.extensions.PsiElementExt`)

Then you can call `ScTemplateDefinition#extendsBlock` method

leonbur replied:
thanks for noticing this, this entire function is unnecessary since I now pass it ScNewTemplateDefinition anyway (it wasn't like that before). so I'll delete it

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+    }
+
+  @tailrec
+  private def findTargetContainer(elem: PsiElement): Either[ScFile, ScTemplateDefinition] = {
```
What about `ScTrait` and `ScEnum`?

leonbur replied:
I'll add `ScTrait`. but `ScEnum` is redundant because it extends `ScClass` anyway

unkarjedy replied:
Could you please at least add a small meaningful test for ScEnum case (and ScTrait)
The fact that `ScEnum` extends `ScClass` is an implementation detail.
We had a lot of refactorings of `ScEnum` in 2023.2/3 and in future there is still a chance the hierarchy will be remade.

leonbur replied:
I have added a test for enum but there's some issue in the testkit where it creates the virtual file for the test from the string it doesn't build the AST correctly and doesn't contain the `ScEnum` in it. manually running the function when intelliJ is running works correctly.

Since the test doesn't create the ScEnum node correctly, the test fails because the created class is generated on the top level instead of inside the enum. please try to run `ScalaAnonymousToInnerTest.scala#testInsideEnum`.
The rest of the tests pass correctly.

unkarjedy replied:
Consider overriding `supportedIn` method in the test class.
Example:
```scala
override protected def supportedIn(version: ScalaVersion): Boolean = version.isScala3
```

leonbur replied:
It worked! thanks

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+    DialogResult(className, renamedVariables) <- showRefactoringDialog(project, anonClass, variables, targetContainer)
+  } yield performRefactoring(project, className, renamedVariables, anonClass, element, targetContainer)
+
+  def parseInitialAnonymousClass(element: ScNewTemplateDefinition): Option[(ScExtendsBlock, Array[VariableData], Either[ScFile, ScTemplateDefinition])] = for {
```
Please mark this method with  `@VisibleForTesting` annotation
(`org.jetbrains.annotations.VisibleForTesting`)

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaMoveAnonymousToInnerDelegate.scala:
```diff
@@ -0,0 +1,35 @@
+    maybeElement.flatMap(element => Option(PsiTreeUtil.getParentOfType(element, classOf[ScNewTemplateDefinition])))
+  }
+
+  val handler = new ScalaAnonymousToInnerHandler
```
I wonder if we need this handler as a class at all.
AFAIS it doesn't contain any state.
Maybe we can make it `ScalaAnonymousToInnerHandler`

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,162 @@
+      case None => None
+    }
+
+  @tailrec
```
(minor) alternative implementation without recursion (in current method)

```scala
  private def findTargetContainer(elem: PsiElement): Either[ScFile, ScTemplateDefinition] = {
    val container = elem.parents.collectFirst {
      case obj: ScObject => Right(obj)
      case clazz: ScClass => Right(clazz)
      case file: ScFile => Left(file) //Scala3 top level definition
    }
    container.get //assuming that there is at least "ScFile" parent, so it's safe to call "get"
  }
```

leonbur replied:
there is no `parents` collection on `PsiElement`. if there were, I would have used that instead.

unkarjedy replied:
It's available in the same extension methods wrapper as `parentOfType`:
`import org.jetbrains.plugins.scala.extensions.PsiElementExt`
Also, IntelliJ suggests to import it (though you need to first type the name)

<img width="500" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/7326e748-715c-49ae-9d15-99ca72cbf91c">

leonbur replied:
oh, IJ didn't suggest it to me at first. I will use your version.
regarding your comment on `container.get`, why wouldn't there be an `ScFile` at the top?

unkarjedy replied:
> why wouldn't there be an ScFile at the top?

There should be.
But it's generally not nice to use `.get`, so with the explicit comment I reassure that we don't expect NPE because at least `ScFile` is expected as one of the parents.

With your implementation, I would add a similar comment to `elem.getParent`

unkarjedy commented on scala/scala-impl/test/org/jetbrains/plugins/scala/lang/refactoring/anonymousToInner/ScalaAnonymousToInnerTest.scala:
```diff
@@ -0,0 +1,152 @@
+
+    val scalaFile = getFile.asInstanceOf[ScalaFile]
+
+    val myElement = scalaFile.findReferenceAt(caretModel.getOffset).getElement
```
`my` prefix is used for java fields.
Just `element` is enough.

unkarjedy replied:
Also this can be shorter:
`val element = scalaFile.findElementAt(caretModel.getOffset)`

unkarjedy commented on scala/scala-impl/test/org/jetbrains/plugins/scala/lang/refactoring/anonymousToInner/ScalaAnonymousToInnerTest.scala:
```diff
@@ -0,0 +1,152 @@
+    val scalaFile = getFile.asInstanceOf[ScalaFile]
+
+    val myElement = scalaFile.findReferenceAt(caretModel.getOffset).getElement
+    val newTemplateDefinition = handler.findParentOfType[ScNewTemplateDefinition](myElement).getOrElse(throw new Exception("expected to find the `new` keyword before the anonymous type"))
```
Too long line, can move the exception to the new line:
```scala
val newTemplateDefinition = handler.findParentOfType[ScNewTemplateDefinition](element).getOrElse {
      throw new Exception("expected to find the `new` keyword before the anonymous type")
    }
    ```

unkarjedy commented:
@leonbur 
Thanks a lot for the contribution!
It's a nice feature which I would also personally use.

azdrojowa123 committed:
sbt: backport code responsible for marking SbtNestedModules as orphan

- insertion of this code will allow the removal of SbtNestedModules when using 2023.3. Without this code, if someone downgrades from 2024 to 2023.3, after the reload, there would be undeleted modules created from SbtNestedModuleData.

azdrojowa123 committed:
sbt: remove dummy test

azdrojowa123 committed:
Revert "sbt: do not use project to reset the model in SbtProjectSettingsControl.scala #SCL-21994 fixed"

This reverts commit adf90e06f86bec8aeb711e0b864b9f511a1cd29c.

vasilmkd committed:
Provide a project constructor parameter when instantiating the global search scope in the sbt build module scope enlarger

#SCL-22013

- The linked issue contains logs with a NullPointerException being thrown for calls to `GlobalSearchScope#getProject`, which can be `null` if no project instance is provided when the scope is constructed.
- It doesn't necessarily mean that the issue is fixed with this, just safer.

vasilmkd committed:
Accept an explicit project function parameter instead on relying on the global search scope to provide a project, which is not guaranteed

#SCL-22013

unkarjedy commented:
@leonbur Please let us know once you are done and are not planning to push any new changes

leonbur commented:
> @leonbur Please let us know once you are done and are not planning to push any new changes

@unkarjedy 
I'm done. do I need to squash?

unkarjedy commented on scala/scala-impl/test/org/jetbrains/plugins/scala/lang/refactoring/anonymousToInner/ScalaAnonymousToInnerTest.scala:
```diff
@@ -0,0 +1,233 @@
+    Assert.assertEquals(expectedText, scalaFile.getText)
+  }
+
+  override protected def supportedIn(version: ScalaVersion): Boolean = version.isScala3
```
Please move it to the beginning of the class, before all the methods in the test class

unkarjedy commented:
> do I need to squash?

@leonbur 
Yes please.
Also, please add a prefix to the commit message: `[refactoring] `

I will then manually pick the change and run it through our internal CI

Commit:
Merge branch 'JetBrains:idea233.x' into idea233.x

leonbur commented:
@unkarjedy well, I did squash my commits but I think I have made a mess because I pulled from upstream during my development that's why there's a merge in the middle.

unkarjedy commented:
I will cherry-pick this change manually and run it using our internal CI.
Please watch https://youtrack.jetbrains.com/issue/SCL-227 for status update
I will include it to 2023.3 Nightly channel (it will be automatically available in 2024.1 as well)

Thanks again for the contribution!

unkarjedy commented:
@leonbur Sorry, I missed one comment, so I reopened the ticket

Commit:
if a parameter is writable, add a comment that the code might not compile and it needs to be fixed

Commit:
not creating a parameter for fields and parameters that are visible from the target container

Commit:
Merge remote-tracking branch 'origin/idea233.x' into idea233.x

leonbur commented:
@unkarjedy please re-review and let me know if I can squash

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,157 @@
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val containsVars = variables.exists(_.element.isVar)
+    val warningComment = if (containsVars)
```
(I will add it as another comment to track it easier)
I've just realised that simply ignoring "write" operations and adding the comment is not enough.
Here is a modified example:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyTrait {
      override def foo(): Unit = {
        println(localVar)
      }
    }

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }
}
```

Notice that `localVar` is not modified inside the anonymous class.
It's only accessed inside it.
But the variable is modified in some other places outside the anonymous class.

The code outputs:
```
0
1
2
```

When you extract the class it looks like this:
```scala
trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyMyTrait(localVar)

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }
  // refactor your code or use scala.runtime.ObjectRef

  class MyMyTrait(localVar: Int) extends MyTrait {
    override def foo(): Unit = {
      println(localVar)
    }
  }
}
```

Now `MyMyTrait` accesses its local copy of the var.
And the output of the program is:
```
0
0
2
```

unkarjedy replied:
It's a simple example, but it illustrates that the refactoring still can break the code silently.
I see two solutions:

1. Ignore `var`s, don't include them as a parameter.
The code can potentially become red, and user will have to handle it manually.
Leave `//todo` comments where needed to mention that we can't handle vars yet and user needs to use `scala.runtime.*Ref` objects manually.

2. Rewrite the code to `*Ref`
E.g. the example from above will look like:
```scala
import scala.runtime.IntRef

trait MyTrait {
  def foo(): Unit
}

object Example {
  def main(args: Array[String]): Unit = {
    val localVar = new IntRef(0)

    val t = new MyMyTrait(localVar)

    t.foo()
    localVar.elem += 1
    t.foo()
    localVar.elem += 1

    println(localVar)
  }

  class MyMyTrait(localVar: IntRef) extends MyTrait {
    override def foo(): Unit = {
      println(localVar.elem)
    }
  }
}
```

The first approach is the easiest.
If you choose it, please create a new YT ticket (and link to https://youtrack.jetbrains.com/issue/SCL-227) describing the limitation and mentioning the 2nd approach.

The second approach is the correct one, but would require some extra effort.
It might be practical to handle it as a separate change.
The refactoring without `var` handling would be already useful.
Some points to consider:
- Mind that there is `scala.runtime.ObjectRef` and multiple `*Ref` classes for primitive types.
- Mind that there are `scala.runtime.*Ref` and `scala.runtime.Volatile*Ref`

leonbur replied:
In option 1 if we don't include the vars in the parameters they can accidentally reference another field with the same name in the target container and therefore compile without errors. extending your example:

```
trait MyTrait {
  def foo(): Unit
}

object Example {

  var localVar = 10
  
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyTrait {
      override def foo(): Unit = {
        println(localVar)
      }
    }

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }
}
```
that prints
```
0
1
2
```


would become:
```trait MyTrait {
  def foo(): Unit
}

object Example {

  var localVar = 10
  
  def main(args: Array[String]): Unit = {
    var localVar = 0

    val t = new MyMyTrait()

    t.foo()
    localVar += 1
    t.foo()
    localVar += 1

    println(localVar)
  }

  class MyMyTrait() extends MyTrait {
    override def foo(): Unit = {
      println(localVar) //todo: refactor your code or use scala.runtime.ObjectRef
    }
  }
}
```
will print
```
10
10
2
```

As a possible quick solution, I could just throw an error to the user if the code contains var references instead of opening the dialog

unkarjedy replied:
> As a possible quick solution, I could just throw an error to the user if the code contains var references instead of opening the dialog

Indeed, a good tradeoff.
Let's do it.
We can show a red error tooltip to the user, like this:
<img width="489" alt="image" src="https://github.com/JetBrains/intellij-scala/assets/3989292/e33d46d8-d8b8-42fe-8c41-f0294f3dc251">
You might use `com.intellij.refactoring.util.CommonRefactoringUtil#showErrorHint`

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,157 @@
+  private def createClass(name: String, anonClass: ScExtendsBlock, variables: Array[ScalaVariableData], project: Project): (ScClass, Option[PsiComment]) = {
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val containsVars = variables.exists(_.element.isVar)
```
Handling of vars should be covered in tests

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,157 @@
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            val containingClass = p.getDeclarationScope
```
Ignoring of containing class fields/parameters should be covered with tests

unkarjedy commented on scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala:
```diff
@@ -0,0 +1,157 @@
+              res = p :: res
+          case r: ScReferencePattern =>
+            val containingClass = r.containingClass
+            if (targetContainer.fold(_ => false, scTemplateDefinition => containingClass != scTemplateDefinition && !containingClass.isAncestorOf(scTemplateDefinition) && !findTargetContainer(r).isLeft))
```
This looks very similar to the code above.
Does this part matter
`!findTargetContainer(r).isLeft)`
?

If so, please, cover it with a separate test case.
It's not that obvious without context, and it's easy to have regressions here.

If it's not imporant difference, then the code with two branches could be refactored to avoid duplication.

unkarjedy replied:
There are multiple issues with the condition:
1. It's quite long and unreadable.
It's better to at least rename `scTemplateDefinition` to just `td`, it's ok for the short lambda parameter

2. It doesn't reflect the semantics.
The condition reads as "add or not add element to collection" without hints about the purpose.
We could fix it by 1. extracting it to var 2. negating the condition
Example
```scala
val isAccessible = targetContainer.fold(_ => true, td => cc == td || cc.isAncestorOf(td) || findTargetContainer(r).isLeft)
if (isAccessible)
  res = r :: res
```
Now the purpose is slightly more clear

Commit:
treating vars from out of the anon class' scope as unsupported. added more tests

Commit:
VisibleForTesting addons

unkarjedy commented:
I will squash it myself and add some extra commits from me.
I will run it through our CI.
